![title](./pic/selecting/bedingte_selektion/1_title.png)

In [18]:
import pandas as pd

In [19]:
df = pd.read_csv('./csv/titanic.csv')
df.head()

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
2   240276   9.6875   NaN        Q  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S

---

Bisher haben wir verschiedene Datenschichten indiziert, indem wir strukturelle Eigenschaften des `DataFrame` selbst verwendet haben. Um jedoch interessante Dinge mit den Daten zu tun, müssen wir oft Fragen stellen, die auf **Bedingungen** basieren.

Hierfür gibt es 4 unterschiedliche Selektoren.

![title](./pic/selecting/bedingte_selektion/2_selektoren.png)

Die Abfrage ist abhängig vom jeweiligen Datentyp der Spalte die abgefragt werden soll. So werden `Strings` in " " abgefragt und `float` / `int` Werte ohne:

In [20]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

Spalte vom Typ `int64`...

In [23]:
df.Age == 47

0      False
1       True
2      False
3      False
4      False
       ...  
413    False
414    False
415    False
416    False
417    False
Name: Age, Length: 418, dtype: bool

...Spalte vom Typ `object` also in diesem Fall `String`

In [22]:
df.Sex == 'female'

0      False
1       True
2      False
3      False
4       True
       ...  
413    False
414     True
415    False
416    False
417    False
Name: Sex, Length: 418, dtype: bool

Diese Operation erzeugt eine Reihe von `Wahr/Falsch`-Booleschen, die auf dem Geschlecht jedes Passagieres basieren. Dieses Ergebnis kann dann innerhalb von `[]` verwendet werden, um die relevanten Daten auszuwählen:

---

## Gleich: `==`

![title](./pic/selecting/bedingte_selektion/3_gleich.png)

<video width="1000" controls src="./pic/selecting/bedingte_selektion/4_gleich_examples.mp4" />

Nehmen wir zum Beispiel an, dass wir uns speziell für Passagiere die Frauen interessieren:

In [7]:
df[df.Sex == 'female']

PassengerId  Survived  Pclass  \
1            893         1       3   
4            896         1       3   
6            898         1       3   
8            900         1       3   
12           904         1       1   
..           ...       ...     ...   
409         1301         1       3   
410         1302         1       3   
411         1303         1       1   
412         1304         1       3   
414         1306         1       1   

                                                Name     Sex   Age  SibSp  \
1                   Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
4       Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1   
6                               Connolly, Miss. Kate  female  30.0      0   
8          Abrahim, Mrs. Joseph (Sophie Halaut Easu)  female  18.0      0   
12     Snyder, Mrs. John Pillsbury (Nelle Stevenson)  female  23.0      1   
..                                               ...     ...   ...    ...   
409                        Peacock, Miss. Treasteall  female   3.0      1   
410                           Naughton, Miss. Hannah  female   NaN      0   
411  Minahan, Mrs. William Edward (Lillian E Thorpe)  female  37.0      1   
412                   Henriksson, Miss. Jenny Lovisa  female  28.0      0   
414                     Oliva y Ocana, Dona. Fermina  female  39.0      0   

     Parch              Ticket      Fare Cabin Embarked  
1        0              363272    7.0000   NaN        S  
4        1             3101298   12.2875   NaN        S  
6        0              330972    7.6292   NaN        Q  
8        0                2657    7.2292   NaN        C  
12       0               21228   82.2667   B45        S  
..     ...                 ...       ...   ...      ...  
409      1  SOTON/O.Q. 3101315   13.7750   NaN        S  
410      0              365237    7.7500   NaN        Q  
411      0               19928   90.0000   C78        Q  
412      0              347086    7.7750   NaN        S  
414      0            PC 17758  108.9000  C105        C  

[152 rows x 12 columns]

<br>

Das resultierende `DataFrame` beinhaltet **152** Zeilen. Das Original hatte **418**. Das bedeutet, dass etwa **36 % der Passagiere Frauen** sind.

<br>

In [8]:
df[df.Pclass == 1]

PassengerId  Survived  Pclass  \
11           903         0       1   
12           904         1       1   
14           906         1       1   
20           912         0       1   
22           914         1       1   
..           ...       ...     ...   
403         1295         0       1   
404         1296         0       1   
407         1299         0       1   
411         1303         1       1   
414         1306         1       1   

                                                  Name     Sex   Age  SibSp  \
11                          Jones, Mr. Charles Cresson    male  46.0      0   
12       Snyder, Mrs. John Pillsbury (Nelle Stevenson)  female  23.0      1   
14   Chaffee, Mrs. Herbert Fuller (Carrie Constance...  female  47.0      1   
20                              Rothschild, Mr. Martin    male  55.0      1   
22                Flegenheim, Mrs. Alfred (Antoinette)  female   NaN      0   
..                                                 ...     ...   ...    ...   
403                             Carrau, Mr. Jose Pedro    male  17.0      0   
404                       Frauenthal, Mr. Isaac Gerald    male  43.0      1   
407                         Widener, Mr. George Dunton    male  50.0      1   
411    Minahan, Mrs. William Edward (Lillian E Thorpe)  female  37.0      1   
414                       Oliva y Ocana, Dona. Fermina  female  39.0      0   

     Parch       Ticket      Fare Cabin Embarked  
11       0          694   26.0000   NaN        S  
12       0        21228   82.2667   B45        S  
14       0  W.E.P. 5734   61.1750   E31        S  
20       0     PC 17603   59.4000   NaN        C  
22       0     PC 17598   31.6833   NaN        S  
..     ...          ...       ...   ...      ...  
403      0       113059   47.1000   NaN        S  
404      0        17765   27.7208   D40        C  
407      1       113503  211.5000   C80        C  
411      0        19928   90.0000   C78        Q  
414      0     PC 17758  108.9000  C105        C  

[107 rows x 12 columns]

---

## Ungleich: `!=`

Wenn also 152 von 418 Passagieren Frauen sind, müssten 266 Passagiere *NICHT*-Frauen sein (also Männer). Das kann mit dem `Ungleich`-Operator überprüft werden.

![title](./pic/selecting/bedingte_selektion/5_ungleich.png)

<video width="1000" controls src="./pic/selecting/bedingte_selektion/6_ungleich_examples.mp4" />

In [9]:
df[df.Sex != 'female']

PassengerId  Survived  Pclass                          Name   Sex   Age  \
0            892         0       3              Kelly, Mr. James  male  34.5   
2            894         0       2     Myles, Mr. Thomas Francis  male  62.0   
3            895         0       3              Wirz, Mr. Albert  male  27.0   
5            897         0       3    Svensson, Mr. Johan Cervin  male  14.0   
7            899         0       2  Caldwell, Mr. Albert Francis  male  26.0   
..           ...       ...     ...                           ...   ...   ...   
407         1299         0       1    Widener, Mr. George Dunton  male  50.0   
413         1305         0       3            Spector, Mr. Woolf  male   NaN   
415         1307         0       3  Saether, Mr. Simon Sivertsen  male  38.5   
416         1308         0       3           Ware, Mr. Frederick  male   NaN   
417         1309         0       3      Peter, Master. Michael J  male   NaN   

     SibSp  Parch              Ticket      Fare Cabin Embarked  
0        0      0              330911    7.8292   NaN        Q  
2        0      0              240276    9.6875   NaN        Q  
3        0      0              315154    8.6625   NaN        S  
5        0      0                7538    9.2250   NaN        S  
7        1      1              248738   29.0000   NaN        S  
..     ...    ...                 ...       ...   ...      ...  
407      1      1              113503  211.5000   C80        C  
413      0      0           A.5. 3236    8.0500   NaN        S  
415      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416      0      0              359309    8.0500   NaN        S  
417      1      1                2668   22.3583   NaN        C  

[266 rows x 12 columns]

In [10]:
df[df.Pclass != 1]

PassengerId  Survived  Pclass  \
0            892         0       3   
1            893         1       3   
2            894         0       2   
3            895         0       3   
4            896         1       3   
..           ...       ...     ...   
412         1304         1       3   
413         1305         0       3   
415         1307         0       3   
416         1308         0       3   
417         1309         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                Kelly, Mr. James    male  34.5      0      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                       Myles, Mr. Thomas Francis    male  62.0      0      0   
3                                Wirz, Mr. Albert    male  27.0      0      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
..                                            ...     ...   ...    ...    ...   
412                Henriksson, Miss. Jenny Lovisa  female  28.0      0      0   
413                            Spector, Mr. Woolf    male   NaN      0      0   
415                  Saether, Mr. Simon Sivertsen    male  38.5      0      0   
416                           Ware, Mr. Frederick    male   NaN      0      0   
417                      Peter, Master. Michael J    male   NaN      1      1   

                 Ticket     Fare Cabin Embarked  
0                330911   7.8292   NaN        Q  
1                363272   7.0000   NaN        S  
2                240276   9.6875   NaN        Q  
3                315154   8.6625   NaN        S  
4               3101298  12.2875   NaN        S  
..                  ...      ...   ...      ...  
412              347086   7.7750   NaN        S  
413           A.5. 3236   8.0500   NaN        S  
415  SOTON/O.Q. 3101262   7.2500   NaN        S  
416              359309   8.0500   NaN        S  
417                2668  22.3583   NaN        C  

[311 rows x 12 columns]

---

## Kleiner oder Kleiner-Gleich: `<` `<=`

Willst du beispielsweise wissen, wieviele Kleinkinder sich an Board befunden haben, kannst du das mit dem Feature Alter abfragen. Dazu lasssen wir uns alle Passagiere mit dem `Kleiner`-Operator ausgeben, die jünger als 3 Jahre sind:

![title](./pic/selecting/bedingte_selektion/7_kleiner.png)

<video width="1000" controls src="./pic/selecting/bedingte_selektion/8_kleiner_examples.mp4" />

In [11]:
df[df['Age'] < 3.0]

PassengerId  Survived  Pclass                                     Name  \
89           981         0       2              Wells, Master. Ralph Lester   
117         1009         1       3          Sandstrom, Miss. Beatrice Irene   
201         1093         0       3  Danbom, Master. Gilbert Sigvard Emanuel   
250         1142         1       2                    West, Miss. Barbara J   
263         1155         1       3             Klasen, Miss. Gertrud Emilia   
281         1173         0       3           Peacock, Master. Alfred Edward   
284         1176         1       3              Rosblom, Miss. Salli Helena   
296         1188         1       2                    Laroche, Miss. Louise   
307         1199         0       3                Aks, Master. Philip Frank   
354         1246         1       3  Dean, Miss. Elizabeth Gladys Millvina""   

        Sex   Age  SibSp  Parch              Ticket     Fare Cabin Embarked  
89     male  2.00      1      1               29103  23.0000   NaN        S  
117  female  1.00      1      1             PP 9549  16.7000    G6        S  
201    male  0.33      0      2              347080  14.4000   NaN        S  
250  female  0.92      1      2          C.A. 34651  27.7500   NaN        S  
263  female  1.00      1      1              350405  12.1833   NaN        S  
281    male  0.75      1      1  SOTON/O.Q. 3101315  13.7750   NaN        S  
284  female  2.00      1      1              370129  20.2125   NaN        S  
296  female  1.00      1      2       SC/Paris 2123  41.5792   NaN        C  
307    male  0.83      0      1              392091   9.3500   NaN        S  
354  female  0.17      1      2           C.A. 2315  20.5750   NaN        S

In [12]:
df[df.SibSp < 1]

PassengerId  Survived  Pclass                            Name     Sex  \
0            892         0       3                Kelly, Mr. James    male   
2            894         0       2       Myles, Mr. Thomas Francis    male   
3            895         0       3                Wirz, Mr. Albert    male   
5            897         0       3      Svensson, Mr. Johan Cervin    male   
6            898         1       3            Connolly, Miss. Kate  female   
..           ...       ...     ...                             ...     ...   
412         1304         1       3  Henriksson, Miss. Jenny Lovisa  female   
413         1305         0       3              Spector, Mr. Woolf    male   
414         1306         1       1    Oliva y Ocana, Dona. Fermina  female   
415         1307         0       3    Saether, Mr. Simon Sivertsen    male   
416         1308         0       3             Ware, Mr. Frederick    male   

      Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0    34.5      0      0              330911    7.8292   NaN        Q  
2    62.0      0      0              240276    9.6875   NaN        Q  
3    27.0      0      0              315154    8.6625   NaN        S  
5    14.0      0      0                7538    9.2250   NaN        S  
6    30.0      0      0              330972    7.6292   NaN        Q  
..    ...    ...    ...                 ...       ...   ...      ...  
412  28.0      0      0              347086    7.7750   NaN        S  
413   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  39.0      0      0            PC 17758  108.9000  C105        C  
415  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416   NaN      0      0              359309    8.0500   NaN        S  

[283 rows x 12 columns]

Bei dieser Form der Abfrage ist das Alter 3 **exklusiv**. Das heißt, Kinder mit 3 Jahren, sind hier ausgeschlossen und werden bei der Abfrage nicht berücksichtigt. Wollen wir jetzt aber ebenfalls Kinder mit dem Alter 3, können wir entweder die Abfrage `alter < 4` bilden, oder aber `alter <= 3`, was das Alter von 3 **inkludiert**:

In [13]:
df[df['Age'] <= 3.0]

PassengerId  Survived  Pclass                                     Name  \
89           981         0       2              Wells, Master. Ralph Lester   
117         1009         1       3          Sandstrom, Miss. Beatrice Irene   
201         1093         0       3  Danbom, Master. Gilbert Sigvard Emanuel   
250         1142         1       2                    West, Miss. Barbara J   
263         1155         1       3             Klasen, Miss. Gertrud Emilia   
281         1173         0       3           Peacock, Master. Alfred Edward   
284         1176         1       3              Rosblom, Miss. Salli Helena   
296         1188         1       2                    Laroche, Miss. Louise   
307         1199         0       3                Aks, Master. Philip Frank   
354         1246         1       3  Dean, Miss. Elizabeth Gladys Millvina""   
409         1301         1       3                Peacock, Miss. Treasteall   

        Sex   Age  SibSp  Parch              Ticket     Fare Cabin Embarked  
89     male  2.00      1      1               29103  23.0000   NaN        S  
117  female  1.00      1      1             PP 9549  16.7000    G6        S  
201    male  0.33      0      2              347080  14.4000   NaN        S  
250  female  0.92      1      2          C.A. 34651  27.7500   NaN        S  
263  female  1.00      1      1              350405  12.1833   NaN        S  
281    male  0.75      1      1  SOTON/O.Q. 3101315  13.7750   NaN        S  
284  female  2.00      1      1              370129  20.2125   NaN        S  
296  female  1.00      1      2       SC/Paris 2123  41.5792   NaN        C  
307    male  0.83      0      1              392091   9.3500   NaN        S  
354  female  0.17      1      2           C.A. 2315  20.5750   NaN        S  
409  female  3.00      1      1  SOTON/O.Q. 3101315  13.7750   NaN        S

Wie du also hier sehen kannst, ist Passagier 409 ein dreijähriges Kind und befindet sich nun auch in dem neuen `DataFrame`

---

## Größer oder Größer-Gleich: `>` `>=`

Wer ist als Familie gereist und wer nicht? Das erfährst du über das **SibSp** Feature. 0 Bedeutet in diesem Fall der Passagier ist alleine gereist. `>`0 dass sich mindestens ein Familienmitglied mit an Board befunden hat.

![title](./pic/selecting/bedingte_selektion/9_groesser.png)

<video width="1000" controls src="./pic/selecting/bedingte_selektion/10_groesser_examples.mp4" />

In [25]:
df[df['SibSp'] > 0]

PassengerId  Survived  Pclass  \
1            893         1       3   
4            896         1       3   
7            899         0       2   
9            901         0       3   
12           904         1       1   
..           ...       ...     ...   
406         1298         0       2   
407         1299         0       1   
409         1301         1       3   
411         1303         1       1   
417         1309         0       3   

                                                Name     Sex   Age  SibSp  \
1                   Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
4       Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1   
7                       Caldwell, Mr. Albert Francis    male  26.0      1   
9                            Davies, Mr. John Samuel    male  21.0      2   
12     Snyder, Mrs. John Pillsbury (Nelle Stevenson)  female  23.0      1   
..                                               ...     ...   ...    ...   
406                        Ware, Mr. William Jeffery    male  23.0      1   
407                       Widener, Mr. George Dunton    male  50.0      1   
409                        Peacock, Miss. Treasteall  female   3.0      1   
411  Minahan, Mrs. William Edward (Lillian E Thorpe)  female  37.0      1   
417                         Peter, Master. Michael J    male   NaN      1   

     Parch              Ticket      Fare Cabin Embarked  
1        0              363272    7.0000   NaN        S  
4        1             3101298   12.2875   NaN        S  
7        1              248738   29.0000   NaN        S  
9        0           A/4 48871   24.1500   NaN        S  
12       0               21228   82.2667   B45        S  
..     ...                 ...       ...   ...      ...  
406      0               28666   10.5000   NaN        S  
407      1              113503  211.5000   C80        C  
409      1  SOTON/O.Q. 3101315   13.7750   NaN        S  
411      0               19928   90.0000   C78        Q  
417      1                2668   22.3583   NaN        C  

[135 rows x 12 columns]

Auch hier kann wieder der `>=`-Operator verwendet werden. Um das selbe Resultat wie oben zu erhalten, kann hierbei die Abfrage `sibsp >= 1` gestellt werden.

In [15]:
df[df['SibSp'] >= 1]

PassengerId  Survived  Pclass  \
1            893         1       3   
4            896         1       3   
7            899         0       2   
9            901         0       3   
12           904         1       1   
..           ...       ...     ...   
406         1298         0       2   
407         1299         0       1   
409         1301         1       3   
411         1303         1       1   
417         1309         0       3   

                                                Name     Sex   Age  SibSp  \
1                   Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
4       Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1   
7                       Caldwell, Mr. Albert Francis    male  26.0      1   
9                            Davies, Mr. John Samuel    male  21.0      2   
12     Snyder, Mrs. John Pillsbury (Nelle Stevenson)  female  23.0      1   
..                                               ...     ...   ...    ...   
406                        Ware, Mr. William Jeffery    male  23.0      1   
407                       Widener, Mr. George Dunton    male  50.0      1   
409                        Peacock, Miss. Treasteall  female   3.0      1   
411  Minahan, Mrs. William Edward (Lillian E Thorpe)  female  37.0      1   
417                         Peter, Master. Michael J    male   NaN      1   

     Parch              Ticket      Fare Cabin Embarked  
1        0              363272    7.0000   NaN        S  
4        1             3101298   12.2875   NaN        S  
7        1              248738   29.0000   NaN        S  
9        0           A/4 48871   24.1500   NaN        S  
12       0               21228   82.2667   B45        S  
..     ...                 ...       ...   ...      ...  
406      0               28666   10.5000   NaN        S  
407      1              113503  211.5000   C80        C  
409      1  SOTON/O.Q. 3101315   13.7750   NaN        S  
411      0               19928   90.0000   C78        Q  
417      1                2668   22.3583   NaN        C  

[135 rows x 12 columns]

---

## Modulo-Operator `%`

Auch wenn er in diesem Szenario eher selten zum Einsatz kommt, will ich der Vollständigkeitshalber auch auf den vorerst letzten Operator eingehen: der `Modulo`-Operator. 

In [26]:
df[df['Age'] % 2 == 0]

PassengerId  Survived  Pclass  \
2            894         0       2   
4            896         1       3   
5            897         0       3   
6            898         1       3   
7            899         0       2   
..           ...       ...     ...   
401         1293         0       2   
402         1294         1       1   
405         1297         0       2   
407         1299         0       1   
412         1304         1       3   

                                             Name     Sex   Age  SibSp  Parch  \
2                       Myles, Mr. Thomas Francis    male  62.0      0      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
5                      Svensson, Mr. Johan Cervin    male  14.0      0      0   
6                            Connolly, Miss. Kate  female  30.0      0      0   
7                    Caldwell, Mr. Albert Francis    male  26.0      1      1   
..                                            ...     ...   ...    ...    ...   
401                               Gale, Mr. Harry    male  38.0      1      0   
402                Gibson, Miss. Dorothy Winifred  female  22.0      0      1   
405  Nourney, Mr. Alfred (Baron von Drachstedt")"    male  20.0      0      0   
407                    Widener, Mr. George Dunton    male  50.0      1      1   
412                Henriksson, Miss. Jenny Lovisa  female  28.0      0      0   

            Ticket      Fare Cabin Embarked  
2           240276    9.6875   NaN        Q  
4          3101298   12.2875   NaN        S  
5             7538    9.2250   NaN        S  
6           330972    7.6292   NaN        Q  
7           248738   29.0000   NaN        S  
..             ...       ...   ...      ...  
401          28664   21.0000   NaN        S  
402         112378   59.4000   NaN        C  
405  SC/PARIS 2166   13.8625   D38        C  
407         113503  211.5000   C80        C  
412         347086    7.7750   NaN        S  

[157 rows x 12 columns]